mnist_cuda_rpc_conv_error/mnist_cuda_rpc_conv.py

```diff
@@ -0,0 +1,124 @@
+        transforms.ToTensor(),
+        transforms.Normalize((0.1307,), (0.3081,))
+    ])
+
+    train_data = datasets.MNIST('./data', train=True, download=True, transform=transform)
+    valid_data = datasets.MNIST('./data', train=False, transform=transform)
+
+    train_dataloader = torch.utils.data.DataLoader(train_data, batch_size=batch_size)
+    valid_dataloader = torch.utils.data.DataLoader(valid_data, batch_size=batch_size)
+
+    input_size = 28 * 28
+    hidden_sizes = [128, 64]
+    output_size = 10
+
+    model = MyRPCPipeline(
+        MyRPCPipelineWrapper(nn.Sequential(
+            nn.Conv2d(1, 32, 3, 1),
+            nn.ReLU(),
+            nn.Conv2d(32, 64, 3, 1),
+            nn.ReLU(),
+            nn.MaxPool2d(2),
+            nn.Dropout(0.25),
+            nn.Flatten(1),
+            nn.Linear(9216, 128),
+            nn.ReLU(),
+            nn.Dropout(0.5),
+            nn.Linear(128, 10),
+        ), "worker1:0", rpc_mode='cuda'),
+    )
+
+    criterion = nn.CrossEntropyLoss()
+    optimizer = DistributedOptimizer(
+        optim.Adam,
+        model.parameter_rrefs(),
+    )
+
+    loaders = {"train": train_dataloader, "valid": valid_dataloader}
+
+    max_epochs = 10
+    accuracy = {"train": [], "valid": []}
+    for epoch in range(max_epochs):
+        print(f"Epoch: {epoch+1}")
+        epoch_correct = 0
+        epoch_all = 0
+        for k, dataloader in loaders.items():
+            for i, (x_batch, y_batch) in enumerate(tqdm(dataloader)):
+                x_batch = x_batch.to(0)
+                y_batch = y_batch.to(0)
+                if k == "train":
+                    model.train()
+                    with dist_autograd.context() as context_id:
+                        outp = model(x_batch)
+                        preds = outp.argmax(-1)
+                        correct = (preds == y_batch).sum()
+                        all = len(y_batch)
+                        epoch_correct += correct.item()
+                        epoch_all += all
+                        loss = criterion(outp, y_batch)
+                        dist_autograd.backward(context_id, [loss])
+                        optimizer.step(context_id)
+                else:
+                    model.eval()
+                    # with torch.no_grad():
+                    outp = model(x_batch)
+                    preds = outp.argmax(-1)
+                    correct = (preds == y_batch).sum()
+                    all = len(y_batch)
+                    epoch_correct += correct.item()
+                    epoch_all += all
+                                  
+            print(f"Loader: {k}. Accuracy: {epoch_correct/epoch_all}")
+            accuracy[k].append(epoch_correct/epoch_all)
+
+
+def run_worker(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '29500'
+    options = rpc.TensorPipeRpcBackendOptions(num_worker_threads=256)
+
+    if rank == 0:
+        options.set_device_map("worker1", {0:0})
```

Comment:
I didn't get this, why we have 0->0? Any reason that rank0 and rank1 need to share `cuda:0`? Is it because you configured `CUDA_VISIBLE_DEVICES` when running these two processes, so their `cuda:0` are actually different devices?

Reply:
Oh I see, is it because with the current code, the entire pipeline lives on `worker1:0`, and `master` loads input data into `cuda:0` as well, so we have 0->0. I assume the purpose of this is to run some initially tests to verify the correctness?